akka-actor/src/main/scala/akka/util/ImmutableIntMap.scala

```diff
@@ -3,139 +3,139 @@
-  val empty: ImmutableIntMap =
-    new ImmutableIntMap(Array.emptyIntArray, Array.empty)
-
-  private final val MaxScanLength = 10
+  final val empty: ImmutableIntMap = new ImmutableIntMap(Array.emptyIntArray, 0)
 }
 
 /**
  * INTERNAL API
- * Specialized Map for primitive `Int` keys to avoid allocations (boxing).
- * Keys and values are backed by arrays and lookup is performed with binary
- * search. It's intended for rather small (<1000) maps.
+ * Specialized Map for primitive `Int` keys and values to avoid allocations (boxing).
+ * Keys and values are encoded consecutively in a single Int array and does copy-on-write with no
+ * structural sharing, it's intended for rather small maps (<1000 elements).
  */
-@InternalApi private[akka] final class ImmutableIntMap private (
-  private val keys: Array[Int], private val values: Array[Int]) {
-
-  final val size: Int = keys.length
+@InternalApi private[akka] final class ImmutableIntMap private (private final val kvs: Array[Int], final val size: Int) {
+
+  private[this] final def indexForKey(key: Int): Int = {
+    // Custom implementation of binary search since we encode key + value in consecutive indicies.
+    // We do the binary search on half the size of the array then project to the full size.
+    // >>> 1 for division by 2: https://research.googleblog.com/2006/06/extra-extra-read-all-about-it-nearly.html
+    @tailrec def find(lo: Int, hi: Int): Int =
+      if (lo <= hi) {
+        val lohi = lo + hi // Since we search in half the array we don't need to div by 2 to find the real index of key
+        val idx = lohi & ~1 // Since keys are in even slots, we get the key idx from lo+hi by removing the lowest bit if set (odd)
+        val k = kvs(idx)
+        if (k == key) idx
+        else if (k < key) find((lohi >>> 1) + 1, hi)
+        else /* if (k > key) */ find(lo, (lohi >>> 1) - 1)
+      } else ~(lo << 1) // same as -((lo*2)+1): Item should be placed, negated to indicate no match
+
+    find(0, size - 1)
+  }
 
   /**
    * Worst case `O(log n)`, allocation free.
+   * Will return Int.MinValue if not found, so beware of storing Int.MinValues
    */
-  def get(key: Int): Int = {
-    val i = Arrays.binarySearch(keys, key)
-    if (i >= 0) values(i)
-    else Int.MinValue // cant use null, cant use OptionVal, other option is to throw an exception...
+  final def get(key: Int): Int = {
+    // same binary search as in `indexforKey` replicated here for performance reasons.
+    @tailrec def find(lo: Int, hi: Int): Int =
+      if (lo <= hi) {
+        val lohi = lo + hi // Since we search in half the array we don't need to div by 2 to find the real index of key
+        val idx = lohi & ~1 // Since keys are in even slots, we get the key idx from lo+hi by removing the lowest bit if set (odd)
+        val k = kvs(idx)
+        if (k == key) kvs(lohi | 1) // lohi, if odd, already points to the value-index, if even, we set the lowest bit to add 1
```

Comment:
While pondering a while during lunch today I realized that addition was not necessary here (idx + 1), but since we already have `lohi` which is either an even number (a key index) or an odd number (a value index 1 larger than the key index) we can simply "conditionally" set the lowest bit to get the value index.